### ISSUE TITLE
Discussion: Guidelines for printing to stdout

### ISSUE THREAD

**Issue by tertsdiepraam** [uutils/coreutils#3093]
TLDR: What should we recommend for printing to stdout?
- `print!`/`println!`?
- `write!`/`writeln!` to locked stdout?
- `write!`/`writeln!` on a `BufWriter`?
- Some other solution?

<hr>

Printing to stdout sounds simple, but Rust (rightfully) exposes a lot of complexity that other languages attempt to hide. We have several reasons to care about this complexity:
- We want to gracefully handle a variety of edge cases (such as https://github.com/uutils/coreutils/issues/2925).
- We want to make the utils as fast or faster than GNU (or at the very least in the same ballpark as GNU) and IO can be a significant bottleneck. This can really make a difference, for instance in this new PR: https://github.com/uutils/coreutils/pull/3092

The `print!`/`println!` macros are relatively unsuitable for both of these cases. It will lock stdout for every call and only does line buffering by default (I believe). Additionally, it will panic on a write error, so handling edge cases becomes difficult (if not impossible).

<details>
<summary>
Click to see the function that `println!` calls
</summary>

Source: https://doc.rust-lang.org/src/std/io/stdio.rs.html

```rust
/// Write `args` to the capture buffer if enabled and possible, or `global_s`
/// otherwise. `label` identifies the stream in a panic message.
///
/// This function is used to print error messages, so it takes extra
/// care to avoid causing a panic when `local_s` is unusable.
/// For instance, if the TLS key for the local stream is
/// already destroyed, or if the local stream is locked by another
/// thread, it will just fall back to the global stream.
///
/// However, if the actual I/O causes an error, this function does panic.
fn print_to<T>(args: fmt::Arguments<'_>, global_s: fn() -> T, label: &str)
where
    T: Write,
{
    if OUTPUT_CAPTURE_USED.load(Ordering::Relaxed)
        && OUTPUT_CAPTURE.try_with(|s| {
            // Note that we completely remove a local sink to write to in case
            // our printing recursively panics/prints, so the recursive
            // panic/print goes to the global sink instead of our local sink.
            s.take().map(|w| {
                let _ = w.lock().unwrap_or_else(|e| e.into_inner()).write_fmt(args);
                s.set(Some(w));
            })
        }) == Ok(Some(()))
    {
        // Succesfully wrote to capture buffer.
        return;
    }

    if let Err(e) = global_s().write_fmt(args) {
        panic!("failed printing to {}: {}", label, e);
    }
}
```

</details>

There are two viable alternatives, both using `Write` on `StdoutLock`, both without `BufWriter`. This gives us much more control with locking, buffering and error handling as `write!` returns a result.  However, this imposes a surprisingly heavy maintenance burden.

The current version of `ls` uses the `BufWriter` approach and illustrates this burden nicely. First, the `BufWriter` is passed to many functions, which is not horrible, but becomes quite annoying. Second, the errors were reported in the wrong order, because they were written to an unbuffered `stderr`, but interleaved with a buffered `stdout`. The output that was generated *before* an error was therefore buffered, then the error was written and only afterwards was the buffered output written to the terminal. Currently, this is fixed by explicitly flushing the output before writing an error. This works, but is not ideal and not something I'd want to see implemented in every util.

> More info on this issue for `ls` can be found in https://github.com/uutils/coreutils/pull/2809 and https://github.com/uutils/coreutils/issues/2785

So I'd like to discuss this and then set our conclusions as guidelines for all utils in the project. Any ideas are welcome! Also, if I got anything wrong with this post please correct me! I would also love to hear how other Rust programs deal with this issue.

**Comment by sylvestre** [comment]
I guess it is still relevant :)

**Comment by karlmcdowall** [comment]
There are a few cases in the `close-stdout.sh` GNU test that fail because println! causes a panic when stdout points to `/dev/full`.

**Comment by Qelxiros** [comment]
First, #5907 makes me think that we should look into the `Write` with `BufWriter` approach for stderr as well. Given that, should we provide macros that abstract away that logic and use them throughout coreutils? I'm thinking of something like `uuio_init!()` that expands to a `static STDOUT: LazyLock<...>` (and similar for stderr), and then `uu(e)print(ln)!` that write to those (I think we can always reference them with `crate::STDOUT` as long as we only call `uuio_init!` at the top of each crate). I'm not sure how this would work with interleaving; we might need to add more macros that allow us to explicitly flush the one we aren't writing to. Is this something that's worth pursuing? I'd be happy to write a PR to uucore if so.

**Comment by oech3** [comment]
No progress?

**Comment by ChrisDryden** [comment]
To replicate the stdbuf functionality, it would require us to have a wrapper helper function that tracks the ENV variables that are associated with that functionality. Apart from that, my understanding is that the stdout/stderr has to take into account the pipe signals to see if there's a broken pipe before writing and then also having a check to see if stdout or stderr are a broken file descriptor. 

I think we have a roadmap to a solution here with the signal reader approach that was implemented here for the pipe signals https://github.com/uutils/coreutils/pull/9657 and with the file descriptors for stdout and stderr: https://github.com/uutils/coreutils/pull/9664 since in theory it means that we are no longer bound by any of the limitations of the rust runtime.

If my research is correct handling the pipe signals and recognizing the broken stdout and stderr file descriptors represent over 15% of the remaining failing GNU tests and > 20 of the issues in the queue, I think it would be worthwhile to have a focused effort on aligning on a plan for this now.

**Comment by ChrisDryden** [comment]
Similar issue with the current stdbuf incompatibility: https://github.com/uutils/coreutils/issues/9066

**Comment by oech3** [comment]
If we made guidelines for std{out,err} at early, we can make chances to achieve new contributors.

**Comment by oech3** [comment]
We have a policy to avoid exit() which is really lage barrier for `println!` removal since tt requires to wrap many functions by Result.
